File: packages/clients/varity-client-js/tests/client.test.ts
/**
 * Varity S3 Client Tests
 */

import { VarityS3Client } from '../src/varity-s3-client';
import {
  PutObjectCommand,
  GetObjectCommand,
  DeleteObjectCommand,
  ListObjectsV2Command,
  HeadObjectCommand,
  CreateBucketCommand,
  DeleteBucketCommand,
  ListBucketsCommand
} from '@aws-sdk/client-s3';

describe('VarityS3Client', () => {
  let client: VarityS3Client;

  beforeAll(() => {
    client = new VarityS3Client({
      endpoint: process.env.VARITY_S3_ENDPOINT || 'http://localhost:3001',
      credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID || 'test-access-key',
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || 'test-secret-key'
      },
      region: 'us-east-1',
      network: 'arbitrum-sepolia',
      storageBackend: 'filecoin-ipfs'
    });
  });

  describe('Client Configuration', () => {
    it('should create client with default configuration', () => {
      const defaultClient = new VarityS3Client();
      expect(defaultClient).toBeInstanceOf(VarityS3Client);

      const config = defaultClient.getConfig();
      expect(config.gatewayType).toBe('s3');
      expect(config.network).toBe('arbitrum-sepolia');
      expect(config.storageBackend).toBe('filecoin-ipfs');
      expect(config.encryptionEnabled).toBe(true);
    });

    it('should create client with custom configuration', () => {
      const customClient = new VarityS3Client({
        gatewayType: 'gcs',
        network: 'arbitrum-one',
        storageBackend: 'filecoin-lighthouse',
        encryptionEnabled: false
      });

      const config = customClient.getConfig();
      expect(config.gatewayType).toBe('gcs');
      expect(config.network).toBe('arbitrum-one');
      expect(config.storageBackend).toBe('filecoin-lighthouse');
      expect(config.encryptionEnabled).toBe(false);
    });
  });

  describe('Bucket Operations', () => {
    const testBucket = 'test-bucket-' + Date.now();

    it('should create bucket', async () => {
      try {
        const response = await client.createBucket({
          Bucket: testBucket
        });
        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(200);
      } catch (error: any) {
        // May fail if gateway not running - that's OK for unit tests
        console.log('Note: Bucket creation requires running gateway');
      }
    }, 10000);

    it('should list buckets', async () => {
      try {
        const response = await client.listBuckets();
        expect(response).toBeDefined();
        expect(Array.isArray(response.Buckets)).toBe(true);
      } catch (error: any) {
        console.log('Note: List buckets requires running gateway');
      }
    }, 10000);

    it('should delete bucket', async () => {
      try {
        const response = await client.deleteBucket({
          Bucket: testBucket
        });
        expect(response).toBeDefined();
      } catch (error: any) {
        console.log('Note: Delete bucket requires running gateway');
      }
    }, 10000);
  });

  describe('Object Operations', () => {
    const testBucket = 'test-bucket';
    const testKey = 'test-file.txt';
    const testContent = 'Hello, Varity! This is a test file.';

    it('should upload object', async () => {
      try {
        const response = await client.putObject({
          Bucket: testBucket,
          Key: testKey,
          Body: testContent,
          ContentType: 'text/plain',
          Metadata: {
            'test-key': 'test-value',
            'upload-timestamp': new Date().toISOString()
          }
        });

        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(200);
        expect(response.ETag).toBeDefined();
      } catch (error: any) {
        console.log('Note: Upload requires running gateway and existing bucket');
      }
    }, 15000);

    it('should get object metadata', async () => {
      try {
        const response = await client.headObject({
          Bucket: testBucket,
          Key: testKey
        });

        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(200);
        expect(response.ContentLength).toBeGreaterThan(0);
        expect(response.ContentType).toBeDefined();
      } catch (error: any) {
        console.log('Note: Head object requires running gateway');
      }
    }, 10000);

    it('should download object', async () => {
      try {
        const response = await client.getObject({
          Bucket: testBucket,
          Key: testKey
        });

        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(200);
        expect(response.Body).toBeDefined();

        // Read body stream
        if (response.Body) {
          const body = await response.Body.transformToString();
          expect(body).toBe(testContent);
        }
      } catch (error: any) {
        console.log('Note: Download requires running gateway');
      }
    }, 10000);

    it('should list objects', async () => {
      try {
        const response = await client.listObjects({
          Bucket: testBucket,
          MaxKeys: 10
        });

        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(200);
        expect(Array.isArray(response.Contents)).toBe(true);
      } catch (error: any) {
        console.log('Note: List objects requires running gateway');
      }
    }, 10000);

    it('should copy object', async () => {
      try {
        const copyKey = 'test-file-copy.txt';
        const response = await client.copyObject({
          Bucket: testBucket,
          Key: copyKey,
          CopySource: `${testBucket}/${testKey}`
        });

        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(200);
      } catch (error: any) {
        console.log('Note: Copy object requires running gateway');
      }
    }, 10000);

    it('should delete object', async () => {
      try {
        const response = await client.deleteObject({
          Bucket: testBucket,
          Key: testKey
        });

        expect(response).toBeDefined();
        expect(response.$metadata.httpStatusCode).toBe(204);
      } catch (error: any) {
        console.log('Note: Delete requires running gateway');
      }
    }, 10000);
  });

  describe('Advanced Features', () => {
    const testBucket = 'test-bucket';

    it('should generate presigned URL for upload', async () => {
      try {
        const command = new PutObjectCommand({
          Bucket: testBucket,
          Key: 'presigned-upload.txt'
        });

        const url = await client.getSignedUrl(command, 3600);
        expect(url).toBeDefined();
        expect(url).toContain('http');
        expect(url).toContain(testBucket);
      } catch (error: any) {
        console.log('Note: Presigned URL generation requires running gateway');
      }
    }, 5000);

    it('should generate presigned URL for download', async () => {
      try {
        const command = new GetObjectCommand({
          Bucket: testBucket,
          Key: 'test-file.txt'
        });

        const url = await client.getSignedUrl(command, 3600);
        expect(url).toBeDefined();
        expect(url).toContain('http');
        expect(url).toContain(testBucket);
      } catch (error: any) {
        console.log('Note: Presigned URL generation requires running gateway');
      }
    }, 5000);

    it('should handle stream upload', async () => {
      try {
        const { Readable } = require('stream');
        const stream = Readable.from(['Hello ', 'streaming ', 'world!']);

        const response = await client.uploadStream(
          testBucket,
          'stream-test.txt',
          stream,
          { 'upload-method': 'stream' }
        );

        expect(response).toBeDefined();
      } catch (error: any) {
        console.log('Note: Stream upload requires running gateway');
      }
    }, 10000);
  });

  describe('Error Handling', () => {
    it('should handle non-existent bucket', async () => {
      try {
        await client.getObject({
          Bucket: 'non-existent-bucket-' + Date.now(),
          Key: 'non-existent.txt'
        });
        // Should not reach here
        expect(true).toBe(false);
      } catch (error: any) {
        // Expected error
        expect(error).toBeDefined();
      }
    }, 10000);

    it('should handle non-existent object', async () => {
      try {
        await client.getObject({
          Bucket: 'test-bucket',
          Key: 'non-existent-' + Date.now() + '.txt'
        });
        // Should not reach here
        expect(true).toBe(false);
      } catch (error: any) {
        // Expected error
        expect(error).toBeDefined();
      }
    }, 10000);
  });
});

File: packages/clients/varity-client-go/varity/client_test.go
package varity

import (
	"bytes"
	"fmt"
	"os"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func getTestConfig() *Config {
	endpoint := os.Getenv("VARITY_S3_ENDPOINT")
	if endpoint == "" {
		endpoint = "http://localhost:3001"
	}

	accessKey := os.Getenv("AWS_ACCESS_KEY_ID")
	if accessKey == "" {
		accessKey = "test-access-key"
	}

	secretKey := os.Getenv("AWS_SECRET_ACCESS_KEY")
	if secretKey == "" {
		secretKey = "test-secret-key"
	}

	return &Config{
		Endpoint:          endpoint,
		AccessKeyID:       accessKey,
		SecretAccessKey:   secretKey,
		Region:            "us-east-1",
		Network:           "arbitrum-sepolia",
		StorageBackend:    "filecoin-ipfs",
		EncryptionEnabled: true,
	}
}

func TestClientConfiguration(t *testing.T) {
	t.Run("DefaultConfiguration", func(t *testing.T) {
		client, err := NewS3Client(&Config{
			AccessKeyID:     "test",
			SecretAccessKey: "test",
		})
		require.NoError(t, err)
		require.NotNil(t, client)

		config := client.GetConfig()
		assert.Equal(t, "arbitrum-sepolia", config["network"])
		assert.Equal(t, "filecoin-ipfs", config["storageBackend"])
		assert.Equal(t, true, config["encryptionEnabled"])
	})

	t.Run("CustomConfiguration", func(t *testing.T) {
		client, err := NewS3Client(&Config{
			Endpoint:          "http://custom-endpoint:9000",
			AccessKeyID:       "custom-key",
			SecretAccessKey:   "custom-secret",
			Network:           "arbitrum-one",
			StorageBackend:    "filecoin-lighthouse",
			EncryptionEnabled: false,
		})
		require.NoError(t, err)

		config := client.GetConfig()
		assert.Equal(t, "http://custom-endpoint:9000", config["endpoint"])
		assert.Equal(t, "arbitrum-one", config["network"])
		assert.Equal(t, "filecoin-lighthouse", config["storageBackend"])
		assert.Equal(t, false, config["encryptionEnabled"])
	})
}

func TestBucketOperations(t *testing.T) {
	client, err := NewS3Client(getTestConfig())
	require.NoError(t, err)

	bucketName := fmt.Sprintf("test-bucket-%d", time.Now().Unix())

	t.Run("CreateBucket", func(t *testing.T) {
		err := client.CreateBucket(bucketName)
		if err != nil {
			t.Logf("Note: Create bucket requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
	})

	t.Run("ListBuckets", func(t *testing.T) {
		buckets, err := client.ListBuckets()
		if err != nil {
			t.Logf("Note: List buckets requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.NotNil(t, buckets)
	})

	t.Run("DeleteBucket", func(t *testing.T) {
		err := client.DeleteBucket(bucketName)
		if err != nil {
			t.Logf("Note: Delete bucket requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
	})
}

func TestObjectOperations(t *testing.T) {
	client, err := NewS3Client(getTestConfig())
	require.NoError(t, err)

	bucket := "test-bucket"
	key := fmt.Sprintf("test-file-%d.txt", time.Now().Unix())
	content := []byte("Hello, Varity! This is a test file.")

	t.Run("PutObject", func(t *testing.T) {
		metadata := map[string]*string{
			"test-key":         aws.String("test-value"),
			"upload-timestamp": aws.String(time.Now().Format(time.RFC3339)),
		}

		err := client.PutObject(bucket, key, content, metadata)
		if err != nil {
			t.Logf("Note: Upload requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		t.Logf("✓ Upload successful: %s", key)
	})

	t.Run("HeadObject", func(t *testing.T) {
		result, err := client.HeadObject(bucket, key)
		if err != nil {
			t.Logf("Note: Head object requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.NotNil(t, result)
		assert.NotNil(t, result.ContentLength)
		t.Logf("✓ Head object successful: %s", key)
	})

	t.Run("GetObject", func(t *testing.T) {
		data, err := client.GetObject(bucket, key)
		if err != nil {
			t.Logf("Note: Download requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.Equal(t, content, data)
		t.Logf("✓ Download successful: %s", key)
	})

	t.Run("ListObjects", func(t *testing.T) {
		objects, err := client.ListObjects(bucket, nil, aws.Int64(10))
		if err != nil {
			t.Logf("Note: List objects requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.NotNil(t, objects)
		t.Log("✓ List objects successful")
	})

	t.Run("CopyObject", func(t *testing.T) {
		copyKey := fmt.Sprintf("test-copy-%d.txt", time.Now().Unix())
		err := client.CopyObject(bucket, key, bucket, copyKey)
		if err != nil {
			t.Logf("Note: Copy object requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		t.Logf("✓ Copy object successful: %s -> %s", key, copyKey)
	})

	t.Run("DeleteObject", func(t *testing.T) {
		err := client.DeleteObject(bucket, key)
		if err != nil {
			t.Logf("Note: Delete requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		t.Logf("✓ Delete object successful: %s", key)
	})
}

func TestAdvancedOperations(t *testing.T) {
	client, err := NewS3Client(getTestConfig())
	require.NoError(t, err)

	bucket := "test-bucket"

	t.Run("PresignedURLs", func(t *testing.T) {
		key := "presigned-test.txt"

		// Presigned GET URL
		getURL, err := client.PresignGetObject(bucket, key, time.Hour)
		if err != nil {
			t.Logf("Note: Presigned URL requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.Contains(t, getURL, "http")
		assert.Contains(t, getURL, bucket)
		t.Logf("✓ Presigned GET URL: %s...", getURL[:50])

		// Presigned PUT URL
		putURL, err := client.PresignPutObject(bucket, key, time.Hour)
		if err != nil {
			t.Logf("Note: Presigned URL requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.Contains(t, putURL, "http")
		t.Logf("✓ Presigned PUT URL: %s...", putURL[:50])
	})

	t.Run("StreamOperations", func(t *testing.T) {
		key := "stream-test.txt"
		content := []byte("Stream content")

		// Upload stream
		reader := bytes.NewReader(content)
		metadata := map[string]*string{
			"upload-method": aws.String("stream"),
		}
		err := client.PutObjectStream(bucket, key, reader, metadata)
		if err != nil {
			t.Logf("Note: Stream upload requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		t.Log("✓ Stream upload successful")

		// Download stream
		var buffer bytes.Buffer
		err = client.GetObjectStream(bucket, key, &buffer)
		if err != nil {
			t.Logf("Note: Stream download requires running gateway: %v", err)
			t.Skip("Gateway not available")
		}
		assert.Equal(t, content, buffer.Bytes())
		t.Log("✓ Stream download successful")
	})
}

func TestErrorHandling(t *testing.T) {
	client, err := NewS3Client(getTestConfig())
	require.NoError(t, err)

	t.Run("NonExistentBucket", func(t *testing.T) {
		_, err := client.GetObject(fmt.Sprintf("nonexistent-%d", time.Now().Unix()), "test.txt")
		assert.Error(t, err)
	})

	t.Run("NonExistentObject", func(t *testing.T) {
		_, err := client.GetObject("test-bucket", fmt.Sprintf("nonexistent-%d.txt", time.Now().Unix()))
		assert.Error(t, err)
	})
}
